django/db/models/query.py

```diff
@@ -979,17 +979,17 @@ def get_or_create(self, defaults=None, **kwargs):
             return self.get(**kwargs), False
         except self.model.DoesNotExist:
             params = self._extract_model_params(defaults, **kwargs)
-            # Try to create an object using passed params.
             try:
                 with transaction.atomic(using=self.db):
                     params = dict(resolve_callables(params))
                     return self.create(**params), True
-            except IntegrityError:
+            except IntegrityError as e:
                 try:
                     return self.get(**kwargs), False
                 except self.model.DoesNotExist:
-                    pass
-                raise
+                    # Re-raise if the object still doesn't exist.
+                    # this wasn't a race condition but an integrity error.
+                    raise e
```

Comment:
I don't think this is the correct solution. Note the test in the original ticket is still failing for me and based off the ticket discussion, I expect the solution to be more involved

Reply:
Have you given any thoughts on how systematically clearing one-to-one relationship cache will break `select_related("parent").get_or_create(parent=parent)`, how this relates to `update_or_create` and if something similar should be done with `defaults` and `create_defaults` [as described in the ticket](https://code.djangoproject.com/ticket/36489#comment:4).

I appreciate that the test are passing but it seems like there are still a few open questions about which approach should be taken?